### PR TITLE
Add signature validation to tls12-download.exe.

### PR DESCRIPTION
**Pull request by BillyONeal**
image

### CODE REVIEW

File: src/tls12-download.c [hunk 1]
```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
...
+    wtd.dwUnionChoice = WTD_CHOICE_FILE;
+    wtd.pFile = &wtfi;
+    wtd.dwStateAction = WTD_STATEACTION_VERIFY;
+    wtd.dwProvFlags = WTD_REVOCATION_CHECK_CHAIN;
+
+    GUID wt_policy_guid = WINTRUST_ACTION_GENERIC_VERIFY_V2;
+
+    LONG trust_validation_result = WinVerifyTrust(INVALID_HANDLE_VALUE, &wt_policy_guid, &wtd);
+    wtd.dwStateAction = WTD_STATEACTION_CLOSE;
+    (void)WinVerifyTrust(INVALID_HANDLE_VALUE, &wt_policy_guid, &wtd);
```
Comment (ras0219-msft) [inline src/tls12-download.c:362]: What does this second identical call to `WinVerifyTrust` do?

File: src/tls12-download.c [hunk 2]
```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
...
+    if (trust_validation_result == 0)
```
Comment (ras0219-msft) [inline src/tls12-download.c:363]:
```suggestion
    if (trust_validation_result == ERROR_SUCCESS)
```

File: src/tls12-download.c [hunk 3]
```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
...
+        write_message(std_out, L" done.\r\n");
+    }
+    else
+    {
+        exit_code = 1;
+        write_message(std_out, L" failed! ");
+        write_hex(std_out, (DWORD)trust_validation_result);
+        FILE_DISPOSITION_INFO fdi;
+        fdi.DeleteFile = TRUE;
+        if (SetFileInformationByHandle(out_file, FileDispositionInfo, &fdi, sizeof(fdi)) == 0)
```
Comment (ras0219-msft) [inline src/tls12-download.c:374]: Would it make sense to pend the delete before downloading and only un-pend the delete once the signature has been validated?

**Review by ras0219-msft — APPROVED** [review]
LGTM, though I did not read the entire WinTrust documentation to validate every parameter.

File: src/tls12-download.c [hunk 4]
```diff
@@ -209,7 +211,8 @@ int __stdcall entry()
         abort_api_failure(std_out, L"GetEnvironmentVariableW");
     }
 
-    const HANDLE out_file = CreateFileW(out_file_path, FILE_WRITE_DATA, 0, 0, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
+    const HANDLE out_file = CreateFileW(
+        out_file_path, FILE_WRITE_DATA | FILE_READ_DATA | DELETE, 0, 0, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
```
Comment (strega-nil-ms) [inline src/tls12-download.c:215]: Could you add comments to the parameters to tell us what they are?

File: src/tls12-download.c [hunk 5]
```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
...
+    write_message(std_out, L" done.\r\nValidating signature...");
+
+    if (SetFilePointer(out_file, 0, 0, FILE_BEGIN) == INVALID_SET_FILE_POINTER)
+    {
+        abort_api_failure(std_out, L"SetFilePointer");
+    }
+
+    DWORD exit_code = 0;
+
+    WINTRUST_FILE_INFO wtfi;
```
Comment (strega-nil-ms) [inline src/tls12-download.c:342]:
```suggestion
    WINTRUST_FILE_INFO wtfi = {0};
```

File: src/tls12-download.c [hunk 6]
```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
...
+    wtfi.cbStruct = sizeof(wtfi);
+    wtfi.pcwszFilePath = out_file_path;
+    wtfi.hFile = out_file;
+    wtfi.pgKnownSubject = 0;
+
+    WINTRUST_DATA wtd;
+    ZeroMemory(&wtd, sizeof(wtd));
```
Comment (strega-nil-ms) [inline src/tls12-download.c:349]:
```suggestion
    WINTRUST_DATA wtd = {0};
```

**Review by strega-nil-ms — APPROVED** [review]

File: src/tls12-download.c (hunk 5 above)
Comment (ras0219-msft) [reply src/tls12-download.c:342]:
```suggestion
    WINTRUST_FILE_INFO wtfi;
    ZeroMemory(&wtfi, sizeof(wtfi));
```

File: src/tls12-download.c (hunk 2 above)
Comment (BillyONeal) [reply src/tls12-download.c:363]: It's documented to return "zero", not ERROR_SUCCESS: https://docs.microsoft.com/en-us/windows/win32/api/wintrust/nf-wintrust-winverifytrust#return-value

Comment (ras0219-msft) [reply src/tls12-download.c:363]: Thanks, comprehension fail on my part. I didn't see that line and looked to the example[1] to see what success looks like (which uses `ERROR_SUCCESS`).

[1] https://docs.microsoft.com/en-us/windows/win32/seccrypto/example-c-program--verifying-the-signature-of-a-pe-file

File: src/tls12-download.c (hunk 3 above)
Comment (BillyONeal) [reply src/tls12-download.c]: Done

**Review by ras0219-msft — APPROVED** [review]

**Pull request merged**